openff/system/components/foyer.py

```diff
@@ -0,0 +1,152 @@
+from copy import copy
+from typing import Any, Dict, Set, Tuple
+
+import parmed as pmd
+from foyer import Forcefield  # type: ignore
+from openff.toolkit.topology import Topology
+
+from openff.system import unit as u
+from openff.system.components.potentials import Potential, PotentialHandler
+from openff.system.components.system import System
+from openff.system.exceptions import MissingParametersError
+
+
+def _copy_params(
+    params: Dict[str, float], *drop_keys: str, param_units: Dict = None
+) -> Dict:
+    params_copy = copy(params)
+    for drop_key in drop_keys:
+        params_copy.pop(drop_key, None)
+    if param_units:
+        for unit_item, units in param_units.items():
+            params_copy[unit_item] = params_copy[unit_item] * units
+    return params_copy
+
+
+def _get_openmm_force_gen(ff: Forcefield, gen_type: type) -> Any:
+    return list(filter(lambda x: isinstance(x, gen_type), ff.getGenerators())).pop()
+
+
+def _topology_from_parmed(structure: pmd.Structure) -> Topology:
+    """ToDo: Create a openFF Topology from a parmed structure"""
+
+
+def from_foyer(structure: pmd.Structure, ff: Forcefield, **kwargs) -> System:
+    """Create an openFF system object from a parmed structure by applying a foyer Forcefield"""
+    system = System()
+    system.topology = _topology_from_parmed(structure)
+    foyer_atom_handler = FoyerAtomTypes()
+    foyer_bond_handler = FoyerBondHandler()
+    system.handlers["FoyerAtomHandler"] = foyer_atom_handler
+    system.handlers["FoyerBondHandler"] = foyer_bond_handler
+
+    return system
+
+
+class FoyerAtomTypes(PotentialHandler):
+
+    name: str = "Atoms"
+    expression: str = "4*epsilon*((sigma/r)**12-(sigma/r)**6)"
+    independent_variables: Set[str] = {"r"}
+    slot_map: Dict[int, str] = dict()  # type: ignore
+    potentials: Dict[str, Potential] = dict()
+
+    def store_matches(
+        self,
+        forcefield: Forcefield,
+        structure: pmd.Structure,
+    ) -> None:
+        """
+        Populate self.slot_map with key-val pairs of slots
+        and unique potential identifiers
+
+        """
+        type_map = forcefield.run_atomtyping(structure)
+        for key, val in type_map.items():
+            self.slot_map[key] = val["atomtype"]
+
+    def store_potentials(self, forcefield) -> None:
+        from simtk.openmm.app.forcefield import NonbondedGenerator  # type: ignore
+
+        non_bonded_forces_gen = _get_openmm_force_gen(forcefield, NonbondedGenerator)
+
+        if non_bonded_forces_gen:
+            nonbonded_params = non_bonded_forces_gen.params.paramsForType
+
```

Comment:
I think the amount of needed here points to how significantly Foyer would benefit from more de-coupling from OpenMM. Going through OpenMM force objects works but should not be necessary - they are hugely complex data structures targeted at performance in simulations. Ideally it should be able to do stuff like

```python3
import foyer

oplsaa = foyer.Forcefield(name="oplsaa")
oplsaa.get_parameters(group="Atoms", key="opls_151")  # arguments should be named differently
# prints something containing values of sigma/espilon/q
oplsaa.get_parameters(group="Bonds", key=["CT", "CT"])
# prints something containing k, length
...
```

Reply:
This can be achieved with https://github.com/mosdef-hub/foyer/pull/358 in the long run, All in all I agree, this feature is not to be implemented here. However, I will be happy to create an issue in foyer and resolve it there.